plugins/c8y_log_plugin/src/main.rs

```diff
@@ -0,0 +1,233 @@
+) -> Result<mqtt_channel::Connection, anyhow::Error> {
+    let mqtt_port = tedge_config.query(MqttPortSetting)?.into();
+    let mqtt_config = mqtt_channel::Config::default()
+        .with_port(mqtt_port)
+        .with_subscriptions(mqtt_channel::TopicFilter::new_unchecked(
+            C8yTopic::SmartRestRequest.as_str(),
+        ));
+
+    let mqtt_client = mqtt_channel::Connection::new(&mqtt_config).await?;
+    Ok(mqtt_client)
+}
+
+pub async fn create_http_client(
+    tedge_config: &TEdgeConfig,
+) -> Result<JwtAuthHttpProxy, anyhow::Error> {
+    let mut http_proxy = JwtAuthHttpProxy::try_new(tedge_config).await?;
+    let () = http_proxy.init().await?;
+    Ok(http_proxy)
+}
+
+fn create_inofity_file_watch_stream(
+    config_file: &Path,
+) -> Result<EventStream<[u8; 1024]>, anyhow::Error> {
+    let buffer = [0; 1024];
+    let mut inotify = Inotify::init().expect("Error while initializing inotify instance");
+
+    inotify
+        .add_watch(&config_file, WatchMask::CLOSE_WRITE)
+        .expect("Failed to add file watch");
+
+    Ok(inotify.event_stream(buffer)?)
+}
+
+async fn run(
+    config_file: &Path,
+    mqtt_client: &mut Connection,
+    http_client: &mut JwtAuthHttpProxy,
+) -> Result<(), anyhow::Error> {
+    let () = handle_dynamic_log_type_update(mqtt_client, config_file).await?;
+
+    let mut inotify_stream = create_inofity_file_watch_stream(config_file)?;
+
+    loop {
+        tokio::select! {
+                message = mqtt_client.received.next() => {
+                if let Some(message) = message {
+                    if let Ok(payload) = message.payload_str() {
+                        let result = match payload.split(',').next().unwrap_or_default() {
+                            "522" => {
+                                // retrieve smartrest object from payload
+                                let smartrest_obj = SmartRestLogRequest::from_smartrest(&payload)?;
+                                handle_logfile_request_operation(
+                                    &smartrest_obj,
+                                    &config_file,
+                                    mqtt_client,
+                                    http_client,
+                                )
+                                .await
+                            }
+                            _ => {
+                                // Ignore operation messages not meant for this plugin
+                                Ok(())
+                            }
+                        };
+
+                        if let Err(err) = result {
+                            error!("Handling of operation: '{}' failed with {}", payload, err);
+                        }
+                    }
+                }
+                else {
+                    // message is None and the connection has been closed
+                    return Ok(());
+                }
+            }
+            Some(event_or_error) = inotify_stream.next() => {
```

Comment:
Just wondering why this whole `inotify` mechanism has to be handled in a select loop like this? Since it's independent from the operation message handling logic couldn't that be run as an independent async task?

Reply:
We can but there are a couple of issues.

- It needs another MQTT client to send the messages to the c8y. Because the `inotify` is just a notification mechanism to notify on some event. Once the event has been raised some operation has to be performed, this involves sending the result ( log types or type of operations that are supported) to the cloud, this needs MQTT.
- Also some of the data structures must be shared between two threads (main and intofiy), then comes the synchronization issues.

So, Using the select! will help here.